app/src/main/java/io/homeassistant/companion/android/nfc/NfcActivity.kt

```diff
@@ -0,0 +1,295 @@
+import android.widget.Toast
+import androidx.appcompat.app.AlertDialog
+import androidx.appcompat.app.AppCompatActivity
+import androidx.recyclerview.widget.LinearLayoutManager
+import io.homeassistant.companion.android.R
+import io.homeassistant.companion.android.common.dagger.GraphComponentAccessor
+import io.homeassistant.companion.android.domain.integration.Entity
+import io.homeassistant.companion.android.domain.integration.IntegrationUseCase
+import io.homeassistant.companion.android.domain.integration.Service
+import io.homeassistant.companion.android.util.UrlHandler
+import io.homeassistant.companion.android.widgets.ServiceFieldBinder
+import io.homeassistant.companion.android.widgets.SingleItemArrayAdapter
+import io.homeassistant.companion.android.widgets.WidgetDynamicFieldAdapter
+import kotlinx.android.synthetic.main.activity_nfc.*
+import kotlinx.coroutines.*
+import javax.inject.Inject
+
+class NfcActivity : AppCompatActivity() {
+    private val TAG: String = "NfcActivity"
+
+    companion object {
+        fun newInstance(context: Context): Intent {
+            return Intent(context, NfcActivity::class.java)
+        }
+    }
+
+    private val mainScope: CoroutineScope = CoroutineScope(Dispatchers.Main + Job())
+
+    @Inject
+    lateinit var integrationUseCase: IntegrationUseCase
+
+    private var services = HashMap<String, Service>()
+    private var entities = HashMap<String, Entity<Any>>()
+    private var dynamicFields = ArrayList<ServiceFieldBinder>()
+    private lateinit var dynamicFieldAdapter: WidgetDynamicFieldAdapter
+
+    private var nfcUrlToWrite: String? = null
+
+    private var mNfcAdapter: NfcAdapter? = null
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_nfc)
+
+        setSupportActionBar(findViewById(R.id.toolbar))
+        supportActionBar?.setDisplayHomeAsUpEnabled(true)
+
+        // Inject components
+        DaggerProviderComponent
+            .builder()
+            .appComponent((application as GraphComponentAccessor).appComponent)
+            .build()
+            .inject(this)
+
+
+        if (NfcAdapter.ACTION_NDEF_DISCOVERED == intent.action) {
+            nfc_setup.visibility = View.GONE
+            val rawMessages = intent.getParcelableArrayExtra(NfcAdapter.EXTRA_NDEF_MESSAGES)
+            val ndefMessage = rawMessages[0] as NdefMessage?
+            mainScope.launch {
+                handleNFCTag(ndefMessage)
+            }
+        } else {
+            nfc_loading.visibility = View.GONE
+            initUi();
+            mNfcAdapter = NfcAdapter.getDefaultAdapter(this)
+        }
+    }
+
+    private suspend fun handleNFCTag(ndefMessage: NdefMessage?) {
+        val url = ndefMessage?.records?.get(0)?.toUri().toString()
+        //val url = "https://www.home-assistant.io/android/nfc/?url=homeassistant://call_service/light.turn_on?entity_id=light.extended_color_light_2"
```

Comment:
I'm personally not a fan of writing urls that contain the service data in it. 

I would much rather have it work like this:

- Each tag gets a unique number
- When a tag is scanned by a phone or device, the ID of the device is also send in.
- Home Assistant is aware of tags and devices and remembers all tags and devices it encounters
- People can automate with a trigger based on tag ID (and optional scan device ID)

That way we don't have to access tags to update automations and it's clear from an overview which ones we have.

I guess we can keep support for this around, but things like custom event are only allowed for admin users and generally we should not offer to users.

Reply:
@balloob Your proposed solution sounds great to me. It has many advantages and is more flexible.

So the key-points to take away here are: 
- keep the support for universal links (but change the url to `https://www.home-assistant.io/nfc/?url=`)
- add support for scanning tags and sending them directly to home-assistant
  - not sure how this can be done? Does the rest api already support something like tags? Maybe someone can give me some pointers how this could be implemented?
  - we still have to write something onto the nfc tags.. otherwise other apps can pick up the nfc tag discovered event as well. Maybe we can create an url that looks something like this `homeassistant://nfc/random-uuid`. Therefore we are still flexible in terms of what will happen when this nfc-tag is scanned.

Reply:
It's worth noting for iOS at least that _only_ universal link style NFC tags can be read in the background.

My proposal for an ID-based one, would be to fire an event like `mobile_app.nfc_tag_read` and pass in everything after` nfc/` in the URL as a parameter to the event.

`https://www.home-assistant.io/nfc/5f0ba733-172f-430d-a7f8-e4ad940c88d7`
-> `mobile_app.nfc_tag_read` + `{"tag": "5f0ba733-172f-430d-a7f8-e4ad940c88d7"}`

Probably worth including as well is device-specific info.

Reply:
@zacwest I noticed the android also only supports `http`, `https` and `app` as a prefix for the schema. Therefore I went with your proposed url style.

With my latest changes, the app supports both nfc tags with an UUID and nfc tags containing the universal app link. 

I tried firing an event to `mobile_app.nfc_tag_read` as well as `mobile_app_nfc_tag_read` however I can't seem to see it in the ha ui. Is that correct? Do I need to configure something else on my ha instance first? 

In my android logcat output I can see the following request:
`{"type":"fire_event","data":{"event_type":"mobile_app.nfc_tag_read","event_data":{"device_name":"HD1903","tag":"669441d9-35ea-4c0f-bdbd-7eabf071dd1d"}}}`

> if you made an NFC reader it could fire it too

you mean a standalone component in ha itself?

Reply:
> you mean a standalone component in ha itself?

Yeah like a USB NFC reader that you could tag into, for example. We could go with `nfc.tag_read` perhaps.

Reply:
> I tried firing an event to `mobile_app.nfc_tag_read` as well as `mobile_app_nfc_tag_read` however I can't seem to see it in the ha ui. Is that correct? Do I need to configure something else on my ha instance first?

In the Developer Tools you can go to Events and listen to them there. Does it not appear if you listen to e.g. `*`?

Reply:
@zacwest Thank you, I thought it would appear as an event in the list on the right side. But if I listen to `*` it did show up, thank you! :)  Just updated the code so that the event is `nfc.tag_read` now.

Reply:
From the discussion above, I think we should abstract _again_! I'm not as confident in it, but how do you feel about `tag.read`?